core/js/src/main/scala/zio/internal/RingBuffer.scala

```diff
@@ -109,6 +111,80 @@ final class RingBuffer[A](override final val capacity: Int) extends MutableConcu
     }
   }
 
+  override final def offerAll(as: Iterable[A]): Iterable[A] = {
+    var curSeq  = 0L
+    var curHead = 0L
+    var curTail = tail
+    var curIdx  = 0
+    var state   = STATE_LOOP
+    val offers  = as.size.toLong
+    var enqHead = 0L
+    var enqTail = 0L
+
+    while (state == STATE_LOOP) {
+      curHead = head
+      curTail = tail
+      val size      = curTail - curHead
+      val available = capacity - size
+      val forQueue  = math.min(offers, available)
+      if (forQueue == 0) {
+        // There are no elements to offer or no space in the queue, terminate
+        // immediately.
+        state = STATE_FULL
+      } else {
+        // We know that space for this many elements has been reserved in the
+        // queue. However, elements in some of these spaces may be in the
+        // process of being dequeued so we need to check for that.
+        enqHead = curTail
+        enqTail = curTail + forQueue
+        var continue = true
+        while (continue & enqHead < enqTail) {
+          curIdx = posToIdx(enqHead, capacity)
+          curSeq = seq(curIdx)
+          if (curSeq != enqHead) {
+            // The element at this spot has not been dequeued yet, or possibly
+            // has been dequeued and another thread has already enqueued a new
+            // element. We need to abort and retry.
+            continue = false
+          }
+          enqHead += 1
+        }
+        // If all elements have been dequeued, we can do compare and swap to
+        // try to reserve the space in the queue.
+        if (continue && tail == curTail) {
+          // We successfully reserved the space in the queue. We can prepare to
+          // enqueue the elements and publish our changes.
+          tail = enqTail
+          enqHead = curTail
+          state = STATE_RESERVED
+        } else {
+          // Another thread beat us to reserving space in the queue. We need to
+          // abort and retry.
+          state = STATE_LOOP
+        }
+      }
+    }
+
+    if (state == STATE_RESERVED) {
+      // We have successfully resserved space in the queue and have exclusive
+      // ownership of each space until we publish our changes. Enqueue the
+      // elements sequentially and publish our changes as we go.
+      val iterator = as.iterator
+      while (enqHead < enqTail) {
+        val a = iterator.next()
+        curIdx = posToIdx(enqHead, capacity)
+        buf(curIdx) = a.asInstanceOf[AnyRef]
```

Comment:
This is tricky logic to manually verify. I suggest we make a stress test, which can bombard a ring buffer with concurrent operations (of the newly-added variety) and verify invariants. This can be a pretty robust way to catch race conditions in concurrent code.

Reply:
Added.